pymatviz/_preprocess_data.py

```diff
@@ -0,0 +1,264 @@
+
+    # Convert to numeric, forcing non-numeric types to NaN
+    all_values = pd.to_numeric(all_values, errors="coerce")
+
+    # Check for NaN
+    has_nan = False
+    if all_values.isna().to_numpy().any():
+        warnings.warn("NaN found in data", stacklevel=2)
+        has_nan = True
+
+    # Check for infinity
+    has_inf = False
+    if np.isinf(all_values).to_numpy().any():
+        warnings.warn("Infinity found in data", stacklevel=2)
+        has_inf = True
+
+    return has_nan, has_inf
+
+
+def get_df_nest_level(df_in: pd.DataFrame, col: str) -> int:
+    """Check for maximum nest level in a DataFrame column.
+
+    Definition of nest level:
+        Level 0 (no list):       "Fe": 1
+        Level 1 (flat list):     "Co": [1, 2]
+        Level 2 (nested lists):  "Ni": [[1, 2], [3, 4], ]
+        ...
+
+    Args:
+        df_in (pd.DataFrame): The DataFrame to check.
+        col (str): Name of the column to check.
+
+    Returns:
+        int: The maximum nest level.
+    """
+    return df_in[col].map(lambda val: np.asarray(val).ndim).max()
+
+
+def replace_missing_and_infinity(
+    df_in: pd.DataFrame,
+    col: str,
+    missing_strategy: Literal["zero", "mean"] = "mean",
+) -> pd.DataFrame:
+    """Replace missing value (NaN) and infinity.
+
+    Infinity would be replaced by vmax(∞) or vmin(-∞).
+    Missing value would be replaced by selected strategy:
+        - zero: replace with zero
+        - mean: replace with mean value
+
+    Args:
+        df_in (DataFrame): DataFrame to process.
+        col (str): Name of the column to process.
+        missing_strategy: missing value replacement strategy.
+    """
+    # Check for NaN and infinity
+    has_nan, has_inf = check_for_missing_inf(df_in, Key.heat_val)
+
+    if not has_nan and not has_inf:
+        return df_in
+
+    # Can only handle nest level 1 at this moment
+    if (has_nan or has_inf) and get_df_nest_level(df_in, col) > 1:
+        raise RuntimeError("Unable to replace NaN and inf for nest level > 1")
+
+    # Get replacement value for missing and infinity
+    values = df_in[col].explode()
+    numeric_values = pd.to_numeric(values, errors="coerce")
+
+    replacement_nan = 0 if missing_strategy == "zero" else numeric_values.mean()
+    replacement_inf_pos = numeric_values[numeric_values != np.inf].max()
+    replacement_inf_neg = numeric_values[numeric_values != -np.inf].min()
+
+    replacements = {
+        np.nan: replacement_nan,
+        np.inf: replacement_inf_pos,
+        -np.inf: replacement_inf_neg,
+    }
+
+    df_in[col] = df_in[col].apply(
+        lambda val: replacements.get(val, val)
+        if isinstance(val, float)
+        else np.array([replacements.get(v, v) for v in val])
+    )
+
+    return df_in
+
+
+def preprocess_ptable_data(data: SupportedDataType) -> pd.DataFrame:
+    """Preprocess input data for ptable plotters, including:
+        - Convert all data types to pd.DataFrame.
+        - Impute missing values.
```

Comment:
can you add one more sentence here to explain what actual steps are taken during data imputation?

Reply:
Done! It seems the `missing_strategy` arg for `preprocess_ptable_data` is forgotten :( Glad we caught this.